File: haskell/dc.hs
{-
Reverese Polish calculator implementation in Haskell
-}
import System.IO
import Control.Monad
import Control.Applicative

type Stack = [Double]

operaters :: [String]
operaters = ["+", "-", "*", "/"]

main :: IO ()
main = do
  line <- readLine
  forM_ [0..(length line - 1)] $ \i -> do
    let n = line !! i
    let s = []
    -- print $ strToDbl $ n
    print $ calOrPush n s

strToDbl :: String -> Double
strToDbl n = read n::Double

push :: Double -> Stack -> Stack
push a xs = a : xs

pop :: Stack -> (Double, Stack)
pop [] = error "Stack underflow"
pop (x : xs) = (x, xs)

calOrPush :: String -> Stack -> Stack
calOrPush str s =
  if str `elem` operaters
     then calc str s
     else push (strToDbl str) s

readLine :: IO [String]
readLine =
  words <$> getLine

calc :: String -> Stack -> Stack
calc w s =
  case w of
    "+" -> (push (n1 + n2) s2)
    "-" -> (push (n1 - n2) s2)
    "*" -> (push (n1 - n2) s2)
    "/" -> (push (n2 / n1) s2)
    _   -> error "Calicurating error"
    where
      n1 = fst $ pop s
      s1 = snd $ pop s
      n2 = fst $ pop s1
      s2 = snd $ pop s1

File: nim/dc.nim
# Forth style reverese Polish calculator, implementation in Nim
import sequtils, strutils, math, strformat
import rdstdin

type
  Stack = seq[float]
  StackUnderflowError = object of Exception

const
  popOperators = ["p", "."]

# 参照渡し
proc calc(stack: var Stack, op: string): bool =
  var tmp: float
  try:
    case op:
      of "+": tmp = stack.pop + stack.pop
      of "*": tmp = stack.pop * stack.pop
      of "-": tmp = stack.pop; tmp = stack.pop - tmp
      of "/": tmp = stack.pop; tmp = stack.pop / tmp
      else: return false
    stack.add(tmp)
    true
  except IndexError:
    raise newException(StackUnderflowError, "stack underflow")

proc stackOperation(stack: var Stack, op: string): bool =
  try:
    if op in popOperators:
      echo stack.pop
      true
    else:
      false
  except IndexError:
    raise newException(StackUnderflowError, "stack underflow")

proc prettyPrint(s: Stack) =
  let pp = s.mapIt(
    if it mod 1.0 == 0.0:
      $int(it)
    else:
      fmt"{it}"
  )
  echo "[" & pp.join(", ") & "]"

proc main() =
  var
    stack: Stack = @[]
    edited: bool
    line: seq[string]

  while true:
    line = readLineFromStdin("> ").split.filterIt(it != "")

    for atom in line:
      try:
        edited = stack.calc(atom) or stack.stackOperation(atom)
        if not edited:
          stack.add(atom.parseFloat)
      except StackUnderflowError:
        echo "Stack underflow!"
      except ValueError:
        echo "Invalid chars!"
        stack = @[]
    prettyPrint stack

try:
  main()
except IOError:
  echo "***Exit***"
  quit 0
